File: 백준/Gold/14502. 연구소/연구소.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;

public class Main {
    static int[][] graph;
    static int n, m;
    static int ans = 0;
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, 1, -1};

    public static void main(String[] args) {
        // 입력 처리
        Scanner scanner = new Scanner(System.in);
        n = scanner.nextInt();
        m = scanner.nextInt();
        graph = new int[n][m];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                graph[i][j] = scanner.nextInt();
            }
        }

        // 벽 설치 및 결과 출력
        makeWall(0);
        System.out.println(ans);
    }

    static void bfs() {
        int[][] graph2 = Arrays.stream(graph).map(int[]::clone).toArray(int[][]::new);
        ArrayDeque<int[]> queue = new ArrayDeque<>();
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (graph[i][j] == 2) {
                    queue.add(new int[]{i, j});
                }
            }
        }
        while (!queue.isEmpty()) {
            int[] current = queue.poll();
            int x = current[0];
            int y = current[1];
            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];
                if (0 <= nx && nx < n && 0 <= ny && ny < m && graph2[nx][ny] == 0) {
                    graph2[nx][ny] = graph2[x][y] + 1;
                    queue.add(new int[]{nx, ny});
                }
            }
        }
        int cnt = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (graph2[i][j] == 0) {
                    cnt++;
                }
            }
        }
        ans = Math.max(ans, cnt);
    }

    static void makeWall(int cnt) {
        if (cnt == 3) {
            bfs();
            return;
        }
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (graph[i][j] == 0) {
                    graph[i][j] = 1;
                    makeWall(cnt + 1);
                    graph[i][j] = 0;
                }
            }
        }
    }
}

File: 백준/Gold/14502. 연구소/연구소.py
from collections import deque
import copy

n, m = map(int, input().split())
graph = [list(map(int, input().split())) for _ in range(n)]
dx = [-1, 1, 0, 0]
dy = [0, 0, 1, -1]
ans = 0


def bfs():
    global ans
    q = deque()
    for i in range(n):
        for j in range(m):
            if graph[i][j] == 2:
                q.append([i, j])
    graph2 = copy.deepcopy(graph)
    while q:
        x, y = q.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx < n and 0 <= ny < m and graph2[nx][ny] == 0:
                graph2[nx][ny] = 2
                q.append([nx, ny])
    cnt = 0
    for i in range(n):
        for j in range(m):
            if graph2[i][j] == 0:
                cnt += 1
    ans = max(ans, cnt)


def makeWall(cnt, x, y):
    if cnt == 3:
        bfs()
        return
    for i in range(x, n):
        if x != i:
            y = 0
        for j in range(y, m):
            if graph[i][j] == 0:
                graph[i][j] = 1
                makeWall(cnt + 1, i, j)
                graph[i][j] = 0


makeWall(0, 0, 0)
print(ans)